Add custom replace rule to markdownlint configuration

OnkarRuikar opened the pull request:
There are many unwanted characters that we've been replacing time to time. For example, https://github.com/mdn/content/pull/16324, https://github.com/mdn/content/pull/16770#issuecomment-1140765087, https://github.com/mdn/content/pull/17834#pullrequestreview-1025728726 are some of the PRs.
Because of some locales or keyboard settings contributors introduce such characters.

We want to handle cases such as:
- curly double quotes `“` to straight quotes `"` #16770
- curly single quotes `‘` and  `’` to straight quote `'`
- double hyphens `--` to m-dash `—` #17833
- three dots `...` to ellipsis `…`
- replace `(en-US/docs` with `(/en-US/docs` to avoid broken links https://github.com/mdn/content/pull/18670#issuecomment-1193056819
- remove trailing spaces. #19866
- and recently https://github.com/mdn/content/pull/18123#issuecomment-1179823977

The PR adds a custom markdownlint rule/plugin [markdownlint-rule-search-replace](https://github.com/OnkarRuikar/markdownlint-rule-search-replace) to address such issues.

With the `search-replace` rule in place we can run the linter like this:
```shell
# to flag the issues
yarn lint:md

# or to fix automatically
yarn fix:md
```
We can provide plain string or regex pattern to search and replace. So many complex cases can be handled.
Refer the rule [documentation](https://github.com/OnkarRuikar/markdownlint-rule-search-replace#using-markdownlintjson-config-file) for more information.

To see the plugin in action refer https://github.com/mdn/content/pull/18210.
***

cc/ @teoli2003 , @nschonni , @caugner

teoli2003 commented:
We have to pay attention to two things:
- exceptions. For example, there is a page that speaks about curly braces, and this shouldn't be replaces
- `for...of`, `for...in`, etc. These 3 dots shouldn't be replaced by an ellipsis.

So I think we should be conservative or have a way to skip pages (for the first case). And we should skip code snippets (areas enclosed with 1 or 3 backticks)

Finally, like always we must clean MDN first, before adding automation. As this was already cleaned not too long ago, this shouldn't be too difficult.

nschonni commented:
https://github.com/mdn/content/pull/12316 probably needs to be addressed first, since the rules only run on a small subset right now

nschonni commented on package.json:
```diff
@@ -17,5 +17,8 @@
     "env-cmd": "10.1.0"
+  },
+  "devDependencies": {
+    "markdownlint-rule-search-replace": "^0.1.0"
```
everything else seems to be in `dependencies` right now, so this should probably move.
Also missing the related yarn.lock changes

OnkarRuikar replied:
I've committed the updated yarn.lock.
***
> everything else seems to be in dependencies right now, so this should probably move.

But in mdn/yari all lint dependencies are in devDependencies:  https://github.com/mdn/yari/blob/main/package.json#L170

OnkarRuikar commented:
> https://github.com/mdn/content/pull/12316 probably needs to be addressed first,

Currently developers refer to only `.markdownlint.json` from mdn/content for the markdown linting.  And they install `markdownlint-cli` globally themselves.
As the linting happens manually at the moment and it is not affecting any other workflows, we can merge this PR to make the feature available. We do not have to block this PR for #12316.

OnkarRuikar commented:
> `for...of`, `for...in`, etc. These 3 dots shouldn't be replaced by an ellipsis.

Now option to skip code inside backticks, ``` ` ```, ```` ``` ```` is available. Use `"skip_code": true` in the lint rule. 
Check the latest code and [documentation](https://github.com/OnkarRuikar/markdownlint-rule-search-replace#using-markdownlintjson-config-file).

OnkarRuikar committed:
Add replace rule to markdownlint task

nschonni commented on package.json:
```diff
@@ -16,6 +16,7 @@
     "cross-env": "7.0.3",
-    "env-cmd": "10.1.0"
+    "env-cmd": "10.1.0",
+    "markdownlint-rule-search-replace": "^1.0.4"
```
Probably want to drop the `^` and re-run Yarn to update the lockfile. Seems to be the convention to pin the Deps here

OnkarRuikar replied:
I agree. Pinning doesn't break things when dependencies release new versions.

OnkarRuikar committed:
Update yarn.lock, commands, and add more rules

OnkarRuikar commented:
All, I've rebased and updated the lock file. Also added new rules as per our recent experiences.

Added new rules to fix 
- internal links that are not relative
- trailing spaces (markdownlinter doesn't flag/fix more than one trailing space)

Using this we've merged following PR 
- #19964

teoli2003 commented:
Can you rebase this PR? There are merge conflicts and warning that I think were fixed elsewhere?

teoli2003 commented:
This PR will flag these legitimate uses of curly braces forever, or I am wrong?
![Capture d’écran 2022-08-26 à 09 56 33](https://user-images.githubusercontent.com/1466293/186852571-cd4bf46e-b8de-4cc3-8654-0d54ebb75334.png)

teoli2003 commented:
Or maybe we should remove this rule from the guide as it will automatically be fixed by Markdownlint. @dipikabh @wbamberg, any thoughts?

OnkarRuikar commented:
> This PR will flag these legitimate uses of curly braces forever, or am I wrong?

Yes.

There are two options to solve this:
- A. Use HTML codes:
  ```
  - **Incorrect**: Please don’t use &ldquo;curly quotes.&rdquo;
  ```
- B. Use magic comments: https://github.com/mdn/content/pull/18009
  ```
  <!-- markdownlint-disable search-replace -->

  - **Incorrect**: Please don’t use “curly quotes.”

  <!-- markdownlint-enable search-replace -->
  ```

teoli2003 commented:
I think we should use HTML escape sequences in these rare cases: see #17988

OnkarRuikar committed:
Add the dependency

OnkarRuikar commented:
cc/ @caugner

teoli2003 commented on yarn.lock:
```diff
@@ -3065,11 +3065,23 @@ markdownlint-cli@0.32.2:
     minimatch "~5.1.0"
     run-con "~1.2.11"
 
+markdownlint-rule-helpers@^0.16.0:
+  version "0.16.0"
+  resolved "https://registry.yarnpkg.com/markdownlint-rule-helpers/-/markdownlint-rule-helpers-0.16.0.tgz#c327f72782bd2b9475127a240508231f0413a25e"
+  integrity sha512-oEacRUVeTJ5D5hW1UYd2qExYI0oELdYK72k1TKGvIeYJIbqQWAz476NAc7LNixSySUhcNl++d02DvX0ccDk9/w==
+
 markdownlint-rule-helpers@~0.17.2:
   version "0.17.2"
   resolved "https://registry.yarnpkg.com/markdownlint-rule-helpers/-/markdownlint-rule-helpers-0.17.2.tgz#64d6e8c66e497e631b0e40cf1cef7ca622a0b654"
   integrity sha512-XaeoW2NYSlWxMCZM2B3H7YTG6nlaLfkEZWMBhr4hSPlq9MuY2sy83+Xr89jXOqZMZYjvi5nBCGoFh7hHoPKZmA==
 
```
It looks like we need 2 different versions of `markdownlint-rule-helpers` now. Is this a problem, @caugner?

Or maybe `markdownlint-rule-search-replace` can be made to use `~0.17.2`, too, @OnkarRuikar ?

OnkarRuikar replied:
Good catch @teoli2003! I've updated the dependency version to `~0.17.2` in the plugin.
I think I need to setup the dependabot there.

teoli2003 replied:
Thank you!

OnkarRuikar committed:
bump dependancy version

nschonni commented:
I'm not sure if https://github.com/DavidAnson/markdownlint/blob/main/README.md#optionscustomrules is available in the config without migrating to the the CLI2 https://github.com/DavidAnson/markdownlint-cli2#markdownlint-cli2jsonc. Reason I haven't switched to the CLI2 is the lack of a `--fix` flag on the commandline, so you need a whole other config object (that could inherit from a base), which seems awkward.

I really like the quote rules and the URL ones, I'm a little -0 on the elipsis and m-dash ones.

OnkarRuikar committed:
deleberate error for testing

OnkarRuikar commented:
> I'm not sure if [DavidAnson/markdownlint@main/README.md#optionscustomrules](https://github.com/DavidAnson/markdownlint/blob/main/README.md?rgh-link-date=2022-08-31T14%3A45%3A16Z#optionscustomrules) is available in the config...

Actually, we are using https://github.com/igorshubovych/markdownlint-cli project for our linting and not https://github.com/DavidAnson/markdownlint (not directly).

And `markdownlint-cli` does [have option](https://github.com/igorshubovych/markdownlint-cli#usage) (`-r`) to specify custom rule package:
```
markdownlint -r markdownlint-rule-search-replace --fix "mymdfile.md"
```
***

The PR does work. I've deliberately committed the typos to the PR in  [deleberate error for testing](https://github.com/mdn/content/pull/17988/commits/de3c3a1dc7d71806ddc82599f1587e46f276f0c7) commit.
And it has flagged the errors as expected:
![search_replace_demo](https://user-images.githubusercontent.com/87750369/187844157-cf9873f1-659a-41ba-aca2-8c174387982e.png)

* The fix option also works: https://github.com/mdn/content/pull/20194

OnkarRuikar committed:
revert the test commit

dipikabh commented:
> Or maybe we should remove this rule from the guide as it will automatically be fixed by Markdownlint. @dipikabh @wbamberg, any thoughts?

Yes, I think we can remove that rule from the guide (just like we've done for the things Prettier automatically fixes).

teoli2003 reviewed (approved):
LGTM, but let's not merge this during the weekend as we are not here to deal with unexpected problems.
